osu.Desktop/Updater/AppImageUpdateManager.cs

```diff
@@ -0,0 +1,564 @@
+                    UseShellExecute = false,
+                    RedirectStandardError = true,
+                    RedirectStandardOutput = true,
+                };
+
+                return info;
+            }
+            public enum States
+            {
+                NOUPDATESAVAILABLE,
+                UPDATESAVAILABLE,
+                DOWNLOADING,
+                VERIFYING,
+                COMPLETED,
+                CANCELLED
+            }
+            private static AppImageUpdateTool instance;
+            public static AppImageUpdateTool Instance
+            {
+                get
+                {
+                    if (instance is default(AppImageUpdateTool))
+                    {
+                        instance = new AppImageUpdateTool();
+                    }
+                    return instance;
+                }
+            }
+            private States state;
+            public static States State
+            {
+                get
+                {
+                    if (Instance.state is default(States))
+                    {
+                        Instance.state = Instance.hasUpdates() ?
+                            States.UPDATESAVAILABLE :
+                            States.NOUPDATESAVAILABLE;
+                    }
+                    return Instance.state;
+                }
+                private set
+                {
+                    Instance.state = value;
+                }
+            }
+            /// <summary>
+            /// Absolute path to AppImage file (with symlinks resolved)<para />
+            /// See: https://docs.appimage.org/packaging-guide/environment-variables.html
+            /// </summary>
+            public static string AppImagePath
+            {
+                get
+                {
+                    return IsAppImage ?
+                            // TODO: verify whether environment variable APPIMAGE is set in the deployed build
+                            Environment.GetEnvironmentVariable("APPIMAGE") :
+                            //Assume osu.AppImage otherwise (use this for debugging a test image)
+                            $"{RuntimeInfo.StartupDirectory}osu.AppImage";
+                }
+            }
+            /// <summary>
+            /// Checks if osu was launched as an AppImage
+            /// </summary>
+            public static bool IsAppImage
+            {
+                get
+                {
+                    return Environment.GetEnvironmentVariable("APPIMAGE") is string;
+                }
+            }
+            private bool hasUpdates()
+            {
+                try
+                {
+                    using (var process = new Process())
+                    {
+                        process.StartInfo = StartInfo(ARGS.CHECK);
+                        process.Start();
+
+                        process.WaitForExit();
+                        return process.ExitCode == 1;
```

Comment:
The tool finds updates if it returns... 1? Isn't that a failure code?

Reply:
[wiki](https://github.com/AppImage/AppImageUpdate/wiki/Self-updating-AppImages#using-the-existing-appimageupdate-user-interfaces-to-perform-updates)
> Both AppImageUpdate and appimageupdatetool provide an easy to use update check. Just call those apps with the -j flag, and watch the exit code. If it exits with return value 0, the update check succeeded, and there is no newer version available. Status 1 means that the update check succeeded, and an update is available. Any other return code can be considered an error while trying to check for changes.

The tool also writes to STDERR when you using the help or version flag and some messages during updating which I did nor parse.